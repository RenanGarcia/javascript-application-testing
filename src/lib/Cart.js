const carrinho = {
  total: 0,
  itens: [],
  addProduct: adicionarItem,
  removeProduct: removerItem,
  summary: resumoCarrinho,
  checkout: finalizaPedido,
};

function calculaTotal() {
  return carrinho.itens.reduce((acc, item) => acc + item.preco, 0);
}

function adicionarItem(item) {
  carrinho.itens.push(item);
  carrinho.total = calculaTotal();
}

function removerItem(itemParaRemover) {
  carrinho.itens = carrinho.itens.filter(
    (item) => item.nome !== itemParaRemover.nome
  );
  carrinho.total = calculaTotal();
}

function resumoCarrinho() {
  return {
    total: carrinho.total,
    itemsQuantity: carrinho.itens.length,
  };
}

function finalizaPedido() {
  const sumary = resumoCarrinho();
  carrinho.itens = [];
  carrinho.total = 0;
  return sumary;
}

export default carrinho;
