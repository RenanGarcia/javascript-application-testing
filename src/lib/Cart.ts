import find from 'lodash/find';
import remove from 'lodash/remove';
import matches from 'lodash/matches';
import Dinero from 'dinero.js';

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

interface CartProduct {
  title: string;
  price: number;
}
interface CartItem {
  product: CartProduct;
  quantity: number;
}

export default class Cart {
  items = [];

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc.add(Money({ amount: item.quantity * item.product.price }));
    }, Money({ amount: 0 }));
  }

  add(item: CartItem) {
    const itemToFind = { product: item.product };
    if (find(this.items, itemToFind)) {
      remove(this.items, matches(itemToFind));
    }
    this.items.push(item);
  }

  remove(product: CartProduct) {
    remove(this.items, matches({ product }));
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;
    return { total, items };
  }

  checkout() {
    const { total, items } = this.summary();
    this.items = [];
    return { total, items };
  }
}
