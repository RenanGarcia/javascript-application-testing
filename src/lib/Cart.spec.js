import Cart from './Cart';

describe('Cart', () => {
  let cart;
  const product = {
    title: 'Adidas running',
    price: 35388, // 353.88 | R$ 353,88
  };

  const product2 = {
    title: 'Skate',
    price: 41872, // 418.72 | R$ 418,72
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2, // 70776
      };

      cart.add(item);
      expect(cart.getTotal().getAmount()).toBe(
        item.quantity * item.product.price
      );
    });

    it('should only consider the last reported quantity of a product', () => {
      const items = [
        {
          product,
          quantity: 2,
        },
        {
          product,
          quantity: 1,
        },
      ];

      items.forEach((item) => {
        cart.add(item);
      });

      expect(cart.getTotal().getAmount()).toBe(
        items[1].quantity * product.price
      );
    });

    it('should update total when a product gets included and then removed', () => {
      const items = [
        {
          product,
          quantity: 2,
        },
        {
          product: product2,
          quantity: 1,
        },
      ];

      items.forEach((item) => {
        cart.add(item);
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(product2.price);
    });
  });
  describe('checkout()', () => {
    it('should return an onbject with the total and the list of items ', () => {
      cart.add({
        product: product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      });

      cart.checkout();
      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });
});
