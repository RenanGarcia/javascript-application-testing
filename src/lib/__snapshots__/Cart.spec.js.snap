// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an onbject with the total and the list of items  1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Skate",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
